homeassistant/helpers/config_validation.py

```diff
@@ -770,25 +770,25 @@ def validator(config: dict) -> dict:
 
 
 def key_value_schemas(
-    key: str, value_schemas: dict[str, vol.Schema]
-) -> Callable[[Any], dict[str, Any]]:
+    key: str, value_schemas: dict[Hashable, vol.Schema]
+) -> Callable[[Any], dict[Hashable, Any]]:
     """Create a validator that validates based on a value for specific key.
 
     This gives better error messages.
     """
 
-    def key_value_validator(value: Any) -> dict[str, Any]:
+    def key_value_validator(value: Any) -> dict[Hashable, Any]:
         if not isinstance(value, dict):
             raise vol.Invalid("Expected a dictionary")
 
         key_value = value.get(key)
 
-        if key_value not in value_schemas:
-            raise vol.Invalid(
-                f"Unexpected value for {key}: '{key_value}'. Expected {', '.join(value_schemas)}"
-            )
+        if isinstance(key_value, Hashable) and key_value in value_schemas:
+            return cast(Dict[Hashable, Any], value_schemas[key_value](value))
```

Comment:
I ran into what I think is a bug in PyLint here. PyLint only allows `dict` to be subscriptable in a typing setting and it doesn't consider the first arg a typing arg and so raises the following error:

```
************* Module homeassistant.helpers.config_validation
homeassistant/helpers/config_validation.py:787:24: E1136: Value 'dict' is unsubscriptable (unsubscriptable-object)
```

Replacing `dict` with `Dict` resolves the error for pylint.

CC @cdce8p

Reply:
That isn't a bug. Since `cast` is runtime evaluated, `dict[..]` would require Python 3.9
In typing context works because of postponed evaluation of type annotations (`from __future__ import annotations`).